### PR TITLE
Using Agent commands to publish test results from PTR V1 and VsTest V1

### PR DESCRIPTION
**Pull request by navin22**
Using agent commands to publish test results from PTR V1 and VSTEST V1.

Legacy tasks use the Powershell commandlets to publish test results, and these commandlets contact TFS and then TCM will be called. In order to move them to TCM directly will be using Agent commands as commandlets code has been deleted.

Using the pipeline variable to switch to old code path in case for any fallback.

Testing Done:
1. Verified if the fallback using `UsePowerShellScripts" is working.
2. Verified VsTestV1.
3. Verified PTR V1 with TRX, Junit, Nunit, Xunit.
[Failing for the non trx formats, only in agent version 2.138, .1,.2,.3,.4 and fixed in 2.138.5]
4. Using Folder names with space.
5. Using wildcards

### CODE REVIEW

File: Tasks/PublishTestResultsV1/PublishTestResults.ps1 [hunk 1]
```diff
@@ -75,30 +81,46 @@ try
...
 	    }
 	    else
 	    {
 		    if($publishRunLevelAttachmentsExists)
 		    {
-			    Publish-TestResults -TestRunner $testRunner -TestResultsFiles $matchingTestResultsFiles -MergeResults $mergeResults -Platform $platform -Configuration $configuration -Context $distributedTaskContext -PublishRunLevelAttachments $publishResultsOption -RunTitle $testRunTitle
+                if($switchToPowerShell) {
+                    Publish-TestResults -TestRunner $testRunner -TestResultsFiles $matchingTestResultsFiles -MergeResults $mergeResults -Platform $platform -Configuration $configuration -Context $distributedTaskContext -PublishRunLevelAttachments $publishResultsOption -RunTitle $testRunTitle
+                } else {
+                    Write-Host "##vso[results.publish type=VSTest;mergeResults=$mergeResults;publishRunAttachments=$publishResultsOption;resultFiles=$matchingTestResultsFiles;platform=$platform;configuration=$configuration;testRunSystem=$testRunner;runTitle=$testRunTitle]"
```
Comment (nigurr) [inline Tasks/PublishTestResultsV1/PublishTestResults.ps1]: fix indention

**Comment by navin22** [comment]
@nigurr I need to do version check as this will break the users who uses agent version 2.138, .1,.2,.3,.4 and it's fixed in 2.138.5.
We had introduced one NRE in 2.138 and hot fixed in 2.138.5

I see in Kusto lot of them are still in 2.138.4, Planning to compare in my code if my version is not 2.138..2.138.4 then use the agent?

I checked things are working fine if agent version >= 2.138.5 and <= 2.136

File: Tasks/PublishTestResultsV1/PublishTestResults.ps1 [hunk 2]
```diff
@@ -75,30 +81,46 @@ try
...
 			    {
 			        Write-Warning "Update the build agent to be able to opt out of test run attachment upload." 
-			    }
-			    Publish-TestResults -TestRunner $testRunner -TestResultsFiles $matchingTestResultsFiles -MergeResults $mergeResults -Platform $platform -Configuration $configuration -Context $distributedTaskContext
+                }
+                if($switchToPowerShell) {
+                    Publish-TestResults -TestRunner $testRunner -TestResultsFiles $matchingTestResultsFiles -MergeResults $mergeResults -Platform $platform -Configuration $configuration -Context $distributedTaskContext
+                } else {
+                    Write-Host "##vso[results.publish type=VSTest;mergeResults=$mergeResults;resultFiles=$matchingTestResultsFiles;platform=$platform;configuration=$configuration;testRunSystem=$testRunner]"
+                }
```
[the remark this answers is not in the document]
Comment (navin22) [reply Tasks/PublishTestResultsV1/PublishTestResults.ps1]: It was copy mistake, fixed it.

File: Tasks/VsTestV1/VSTest.ps1 [hunk 3]
```diff
@@ -271,15 +287,23 @@ finally
             
                     if($publishRunLevelAttachmentsExists)		
                     {
-                        Publish-TestResults -Context $distributedTaskContext -TestResultsFiles $resultFiles -TestRunner "VSTest" -Platform $platform -Configuration $configuration -PublishRunLevelAttachments $publishResultsOption
+                        if($switchToPowerShell -ieq "true") {
+                            Publish-TestResults -Context $distributedTaskContext -TestResultsFiles $resultFiles -TestRunner "VSTest" -Platform $platform -Configuration $configuration -PublishRunLevelAttachments $publishResultsOption
+                        } else {
+                            Write-Host "##vso[results.publish type=VSTest;publishRunAttachments=$publishResultsOption;resultFiles=$resultFiles;platform=$platform;configuration=$configuration;testRunSystem=VSTest;]"
```
[the remark this answers is not in the document]
Comment (navin22) [reply Tasks/VsTestV1/VSTest.ps1]: no in vstest we don't need. The bug was only for the non trx result formats.

File: Tasks/PublishTestResultsV1/PublishTestResults.ps1 [hunk 4]
```diff
@@ -61,46 +61,86 @@ try
...
+        try {
+            $agentVersion = [version]$agentVersionVar;
+        }
+        catch {
+            Write-Verbose "Invalid Agent version : $agentVersionVar";
+        }
+
+        # There's a Null reference exception introduced in the agent 2.138 until 2.138.4 and hot fixed in 2.138.5.
+        # Hence falling back to Scripts to pulblish test results. If Agent version falls between 2.138 .. 2.138.4
+        if(!$agentVersion -or (($agentVersion -ge [version]"2.138") -and ($agentVersion -le [version]"2.138.4"))) {
```
[the remark this answers is not in the document]
Comment (navin22) [reply Tasks/PublishTestResultsV1/PublishTestResults.ps1]: I want to support the agent version if they are agent version less than or equal to 2.136 also. They don't have any bug.

File: Tasks/PublishTestResultsV1/PublishTestResults.ps1 [hunk 5]
```diff
@@ -61,46 +61,86 @@ try
...
+                    Write-Host "##vso[results.publish type=$testRunner;mergeResults=$mergeResults;publishRunAttachments=$publishResultsOption;resultFiles=$matchingTestResultsFiles;platform=$platform;configuration=$configuration;testRunSystem=$testRunSystem;runTitle=$testRunTitle;]"
+                }
+            }
+            else 
+            {
+                if(!$publishResultsOption)
+                {
+                    Write-Warning "Update the build agent to be able to opt out of test run attachment upload." 
+                }
+                if($switchToPowerShell-ieq "true") {
```
Comment (nigurr) [inline Tasks/PublishTestResultsV1/PublishTestResults.ps1]: this indetion is not fixed yet. -ieq should be spaced

**Pull request merged**